script.artistslideshow/resources/plugins/fanarttv.py

```diff
@@ -0,0 +1,136 @@
+        if img_params.get( 'clientapikey', False ):
+            self.HASCLIENTKEY = True
+            url_params['client_key'] = img_params.get( 'clientapikey', '' )
+            if img_params.get( 'donate' ) == 'true':
+                self.HASDONATION = True
+                self.CACHEEXPIRE['low'] = self.CACHEEXPIREWITHDONATION
+                self.CACHEEXPIRE['high'] = self.CACHEEXPIREWITHDONATION
+            else:
+                self.CACHEEXPIRE['low'] = self.CACHEEXPIREWITHCLIENTKEY
+                self.CACHEEXPIRE['high'] = self.CACHEEXPIREWITHCLIENTKEY            
+        json_data = self._get_data( filepath, cachefilepath, url, url_params )
+        if json_data:
+            image_list = json_data.get( 'artistbackground', [] )
+            if img_params.get( 'getall', 'false' ) == 'true':
+                image_list.extend( json_data.get( 'artistthumb', [] ) )
+            for image in image_list:
+                url = image.get( 'url', '' )
+                if url:
+                    images.append( url )
+        return images, self.loglines
+
+
+    def _get_cache_time( self, cachefilepath ):
+        rawdata = ''
+        self.loglines.append( 'getting the cache timeout information for fanarttv' )
+        exists, cloglines = checkPath( cachefilepath, False )
+        self.loglines.extend( cloglines )
+        if exists:
+            success = True
+        else:
+            success = self._put_cache_time( cachefilepath )
+        if success:
+            rloglines, rawdata = readFile( cachefilepath ) 
+            self.loglines.extend( rloglines )
+        try:
+            cachetime = int( rawdata )
+        except ValueError:
+            cachetime = 0
+        # this is to honor donation or client key cache time immediately instead of after old cache expires
+        if self.HASDONATION and cachetime > self.CACHEEXPIREWITHDONATION:
+            return self.CACHEEXPIREWITHDONATION
+        elif self.HASCLIENTKEY and cachetime > self.CACHEEXPIREWITHCLIENTKEY:
+            return self.CACHEEXPIREWITHCLIENTKEY
+        else:
+            return cachetime
+
+
+    def _get_data( self, filepath, cachefilepath, url, url_params ):
+        json_data = ''
+        if self._update_cache( filepath, cachefilepath ):
+            success, uloglines, json_data = self.JSONURL.Get( url, params=url_params )
+            self.loglines.extend( uloglines )
+            if success:
+                success, wloglines = writeFile( py2_encode( _json.dumps( json_data ) ), filepath )
+                self.loglines.extend( wloglines )
+        exists, cloglines = checkPath( filepath, False )
+        self.loglines.extend( cloglines )
+        if exists:
+            rloglines, rawdata = readFile( filepath )
+            self.loglines.extend( rloglines )
+            try:
+                json_data = _json.loads( rawdata )
+            except ValueError:
+                success, dloglines = deleteFile( filepath )
```

Comment:
Not sure this is allowed

Reply:
I'm not sure to what you're referring.  This piece of code has existed in AS for quite some time and has previously been allowed in the repo.  But if you can be more specific I can change it.

Reply:
No problem for the script. It is just a concern that this creates an attack vector for other scripts to explore. They can just call the script with whatever file they want, or achieve arbitrary code execution.
And such things can be easily missed out in submissions. Not to talk about external scripts that are not in our repo and and miss our review process.

Reply:
I think I see a way I could add some validation to the filepath so that it's at least in the parent directory I'm expecting. I might be able to check the full path too, but I need to look at that more closely, as AS allows users to set an arbitrary directory where the folder if files is kept. I have that in the settings, I just need to see if I can get it from within this part of the script.  Would that help?

I think I'm about to show my level of Python experience (even after seven years of self taught Python) and understanding of Kodi addons, but how would another addon call an arbitrary function within my code?  If I understood the threat vector better I could take a look at all the AS code and try and address it in a future version.

Reply:
In this script the problem does not reflect much as the addon calling into yours would need to add your file to the python path (e.g. sys.path.append). The problem is more evident in script modules that can just be set as a dependency of another addon. Either way, sometimes it is just a question of common-sense and avoid creating the possible hole in the first place. For instance, in this case, the `os.system` call belongs to a method that doesn't even get called. So why keep it in the first place?
I understand you reuse a lot of the same methods across addons and that's the reason the method is there. Running a linter against your code will detect all those unused methods.
IIRC some time ago team kodi was suggesting addon developers to use `codacity` to perform static analysis on the code automatically.

Reply:
Thanks for helping me better understand all this.  Given the larger security concerns with subprocess and os.system, I went ahead and removed that code.  I'll have to go back and look, but it might only be used in one addon anyway.

Thanks also for the Codacy recommendation.  I had setup codacy sometime ago but forgot about it.  I just ran AS through that and fixed a bunch of stuff (including SO MUCH TRAILING WHITESPACE).  Codacy identified some other security issues with ElementTree, so I moved over to defusedxml, and I had some unverified HTTPS connections with requests that I fixed too.  Some of the method versus function stuff I decided to ignore, as there are readability reasons to have some of the code as methods inside a class even though they could be functions.  I also have one except clause that is unbounded, but I can't find any docs on what kind of error messages xmbc.GetInfoLabel returns, so I had to leave it.   Anyway, Codacy now rates the AS code as an "A."  I'll close this request and submit a new pull request with all the updates (and exclude the two files you mentioned earlier).